### DISCUSSION TITLE
What does the metric "shape" refer to in the compute_cell_quality function?

### DISCUSSION THREAD

**Discussion by Ben-AC-Math** [pyvista/pyvista#6143] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
"""
My code is working but I want to know some further details on the quality measures for mesh evaluation.
"""
def compute_metrics_from_vtk(directory):
    print("Starting compute_metrics_from_vtk...")
    
    vtk_files = [f for f in os.listdir(directory) if f.endswith('.vtk') and f.split('.')[0].isdigit()]

    all_metrics = {
        'aspect_ratio': [],
        'shape': [],
        'scaled_jacobian': []
    }
    
    for vtk_file in vtk_files:
        print(f"Processing {vtk_file}...")
        
        mesh = pv.read(os.path.join(directory, vtk_file))

        # Compute various metrics
        for metric in all_metrics.keys():
            quality_measure = metric if metric != 'jacobians' else 'jacobian'
            try:
                computed_quality = mesh.compute_cell_quality(quality_measure=quality_measure)
                if "CellQuality" not in computed_quality.cell_data:
                    print(f"No {metric} data found in {vtk_file}")
                else:
                    all_metrics[metric].extend(computed_quality.cell_data["CellQuality"])
                    print(f"Found {metric} data in {vtk_file} with length {len(computed_quality.cell_data['CellQuality'])}")
            except Exception as e:
                print(f"Error computing {metric} for {vtk_file}: {e}")

    print("Finished processing all vtk files.")
    return all_metrics
```

### Description

Currently I am evaluating cardiac meshes. I am using Aspect Ratio and Scaled Jacobian. I wanted to include Shape Factor. 

My Question is: What does the quality measure "Shape" refer to? 

I can't find any further documentation on the details of the quality measures.

### System Information

```shell
Not experiencing any bugs.
```

**Comment by darikg** [comment]
it doesn't appear to be documented in [vtkMeshQuality](https://vtk.org/doc/nightly/html/classvtkMeshQuality.html) either so we can spelunk around in the code:

[`vtkMeshQuality::TriangleShape`](https://github.com/Kitware/VTK/blob/82442b87ddc5353b169629997a818075b4400130/Filters/Verdict/vtkMeshQuality.cxx#L1346) calls `verdict::tri_shape(3, pc)` where `pc` are the three points of the triangle

and [`verdict::tri_shape(3, pc)`](https://github.com/Kitware/VTK/blob/82442b87ddc5353b169629997a818075b4400130/ThirdParty/verdict/vtkverdict/V_TriMetric.cpp#L536) appears to simply return `1 / condition`, where condition is calculated from [`verdict::tri_condition`](https://github.com/Kitware/VTK/blob/82442b87ddc5353b169629997a818075b4400130/ThirdParty/verdict/vtkverdict/V_TriMetric.cpp#L466).

Squinting at this code, I *think* it's equivalent to the 'shape regularity` measure described [here](https://onlinelibrary.wiley.com/doi/full/10.1111/cgf.14779) (equation 13), referenced to [this paper](https://www.sciencedirect.com/science/article/pii/S0168874X02000707?casa_token=k8KHdZvrTAkAAAAA:Djiloh75sZr9sD7E-587_TiBDb8UzKdHnUtjzlftpP1DWdQMDYEkYFAW3KIu9Zsgter4A5LdWw).

**Reply by user27182** [reply, marked as answer]
To add to this answer:
> ## The shape regularity metric has a value of one for equilateral triangles and approaches zero for triangles with small angles.

So, a triangle with a value of 1 is "good quality", and a value of 0 is "bad quality". This definition is taken directly from the screenshot provided below.

The vtkMeshQuality docs linked above by @darikg states:
> For more information on the triangle quality functions of this class, cf. Pebay & Baker 2003, Analysis of triangle quality measures, Math Comp 72:244. For more information on the quadrangle quality functions of this class, cf. Pebay 2004, Planar Quadrangle Quality Measures, Eng Comp 20:2."

A PDF of the first paper is publicly accessible here:
https://doi.org/10.1090/S0025-5718-03-01485-6

This paper states:
> shape regularity has been independently suggested by [4]

Where [4] is this reference:
> Bank, Randolph E., and R. Kent Smith. “Mesh Smoothing Using a Posteriori Error Estimates.” SIAM Journal on Numerical Analysis 34, no. 3 (1997): 979–97. http://www.jstor.org/stable/2951970.

If we take a look at that 1997 paper, it has a helpful figure and provides some general intuition:

> shape regularity ... is normalized to equal one for an equilateral triangle and to approach zero for triangles with small angles.

![image](https://github.com/pyvista/pyvista/assets/89109579/7639aa74-770e-4fee-8deb-dc9cf1052aab)

**Reply by Ben-AC-Math** [reply]
Thank you for this! I guess this suggests that this metric is not suitable for a tetrahedral meshes then. 

Thanks again for the detailed answer.

**Reply by darikg** [reply]
Perfect. What the C++ code is returning is 

$$
\frac{ 2 \sqrt 3 |t|} {v_{12} \cdot v_{12} + v_{13} \cdot v_{13} - v_{12} \cdot v_{13}}
$$

which if you substitute the cosine law for the third product and $|l_1|^2 = v_{12} \cdot v_{12}$ etc gives you equation 1 in that screenshot

**Reply by user27182** [reply]
> I  guess this suggests that this metric is not suitable for a tetrahedral meshes then.

It does seem like it only applies to triangle meshes. Certainly the actual source code linked by darikg above reinforces this. Though, it's possible the concept can be generalized, e.g. "normalized to 1 for < _insert regular shape here_ > and approaches 0 for small angles."

The value of the computed cell quality should be -1 if it can't be computed. So, I would expect this filter returns -1 for tetrahedral cells. This should confirm that it doesn't apply for that cell type.

**Reply by darikg** [reply]
there's an equivalent [tet_shape](https://github.com/Kitware/VTK/blob/82442b87ddc5353b169629997a818075b4400130/ThirdParty/verdict/vtkverdict/V_TetMetric.cpp#L1242),I think it works the same way

**Reply by Ben-AC-Math** [reply]
![shape_histogram](https://github.com/pyvista/pyvista/assets/67483352/3c1053f3-f334-4ad8-8860-1a51a61462ed)

It gives a full range when plotted on a histogram. So must have an equivalent tet_shape

**Reply by tkoyama010** [reply]
Thanks for awesome discussion. I added link to this discussion in `compute_cell_quality` docstring in #6150 .

**Reply by user27182** [reply]
As a follow-up, I have found that the `'shape'` metric is the most general one, and applies to the most cells. It's valid for 
`HEXAHEDRON`, `PYRAMID`, `QUAD`, `TETRA`, `TRIANGLE`, and `WEDGE` cells. See the analysis here https://github.com/pyvista/pyvista/pull/7306#issuecomment-2725630614
